applications/MultilevelMonteCarloLaplacianApplication/python_scripts/multilevel_montecarlo_laplacian_analysis.py

```diff
@@ -0,0 +1,724 @@
+        if stochastic_pdf["beta_distribution"].Has("beta"):
+            beta = stochastic_pdf["beta_distribution"]["beta"].GetDouble()
+        else:
+            raise Exception('Please define the "beta" for the beta distribution in the .json file')
+        sample = np.random.beta(alpha,beta,number_samples)
+
+    else:
+        raise Exception('Please provide "normal_distribution" or "beta_distribution" in the .json file; at the moment only "normal_distribution" and "beta_distribution" are implemented')
+    return sample
+
+def GenerateBetaSample(alpha,beta):
+    number_samples = 1
+    sample = np.random.beta(alpha,beta,number_samples)
+    return sample
+
+
+def EvaluateQuantityOfInterest(simulation):
+    """here we evaluate the QoI of the problem: int_{domain} SOLUTION(x,y) dx dy
+    we use the midpoint rule to evaluate the integral"""
+    Kratos.CalculateNodalAreaProcess(simulation._GetSolver().main_model_part,2).Execute()
+    Q = 0.0
+    for node in simulation._GetSolver().main_model_part.Nodes:
+        Q = Q + (node.GetSolutionStepValue(Kratos.NODAL_AREA)*node.GetSolutionStepValue(Poisson.SOLUTION))
+        #print("NODAL AREA = ",node.GetSolutionStepValue(Kratos.NODAL_AREA),"NODAL SOLUTION = ",node.GetSolutionStepValue(Poisson.SOLUTION),"CURRENT Q = ",Q)
+    return Q
+
+
+def Nf_law(lev):
+        # refinement strategy:
+        # uniform mesh on level lev with h_lev=(1/N0)*2^(-lev)
+        # i.e. level = 0, h_{lev=0} = 0.25
+        #      level = 1, h_{lev=1} = 0.2/2 = 0.125
+        #      level = 2, h_{lev=2} = 0.2/4 = 0.0625
+        #      ...
+        # I wrote the following:
+        N0 = 4.
+        M  = 2.
+        NFF = (N0*np.power(M,lev))
+        Nf2 = 2*NFF**2
+        # Nobile wrote the following:
+        # N0 = 5.
+        # M  = 2.
+        # NFF = (N0*np.power(M,lev))
+        # Nf2 = NFF**2
+        # NFF is the number of elements on a boundary line
+        # Nf2 is the number of triangular elements in the square domain (approximately, if mesh non uniform)
+        return Nf2
+
+
+def update_onepass_M(sample, old_mean, old_M2, nsam):
+    # update mean and second moment values
+    # M_{2,n} = sum_{i=1}^{n} (x_i - mean(x)_n)^2
+    # M_{2,n} = M_{2,n-1} + (x_n - mean(x)_{n-1}) * (x_n - mean(x)_{n})
+    # s_n^2 = M_{2,n} / (n-1)
+    delta = np.subtract(sample, old_mean)
+    if nsam == 1:
+        new_mean = sample
+        new_M2 = np.zeros(np.size(sample))
+        new_M2 = np.asscalar(new_M2) # do so to have a list of scalars, and not a list of arrays of one element
+    else:
+        new_mean = old_mean + np.divide(delta,nsam)
+        new_M2 = old_M2 + delta*np.subtract(sample,new_mean)
+
+    return new_mean, new_M2
+
+
+def compute_sample_variance_from_M2(M2,nsam):
+    sample_variance = np.divide(M2,np.subtract(nsam,1))
+    return sample_variance
+
+
+def EstimateBayesianVariance(mean,variance,settings_ML,ratesLS,nDoF,nsam,level_local):
```

Comment:
comment what is this (maybe by referring to the literature)

Reply:
btw, are those functions specific to this case, or can they be used for other applications? if so, please put them in a separate file